File: backend/app/schemas/synthesis.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


ThemeStatus = Literal["ai-generated", "researcher-reviewed", "approved", "rejected"]
ReportStatus = Literal["ai-generated", "researcher-reviewed", "approved", "superseded"]


class SessionEvidenceRead(BaseModel):
    id: str
    document_id: str
    document_name: str
    speaker: str
    location: str
    excerpt: str
    relevance: float
    context_result_id: str


class SessionThemeRead(BaseModel):
    id: str
    project_id: str
    session_id: str
    name: str
    summary: str
    status: ThemeStatus
    confidence: float
    source_label: str
    evidence: list[SessionEvidenceRead]


class SessionThemeUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=180)
    summary: str | None = Field(default=None, min_length=1)
    status: ThemeStatus | None = None


class SessionThemeGenerateResponse(BaseModel):
    themes: list[SessionThemeRead]
    message: str


class SessionReportParticipantRead(BaseModel):
    id: str
    name: str
    role: str | None
    organization: str | None
    notes: str | None


class SessionReportItemRead(BaseModel):
    id: str
    type: Literal["requirement", "decision", "action-item", "open-question", "key-insight"]
    title: str
    summary: str
    provenance: str
    evidence: list[SessionEvidenceRead]


class SessionReportRead(BaseModel):
    id: str
    project_id: str
    session_id: str
    status: ReportStatus
    session_title: str
    session_type: str
    session_date: datetime
    duration_minutes: int | None
    participants: list[SessionReportParticipantRead]
    executive_summary: str
    items: list[SessionReportItemRead]
    detailed_notes: str
    generated_at: datetime


class SessionReportGenerateResponse(BaseModel):
    report: SessionReportRead
    message: str


class SessionReportUpdate(BaseModel):
    status: ReportStatus


class SessionCitationRead(BaseModel):
    id: str
    document_id: str
    document_name: str
    speaker: str
    location: str
    excerpt: str
    context_result_id: str


class SessionConversationTurnRead(BaseModel):
    id: str
    role: Literal["researcher", "assistant"]
    content: str
    citations: list[SessionCitationRead]
    created_at: datetime


class SessionConversationRead(BaseModel):
    id: str
    project_id: str
    session_id: str
    status: Literal["saved", "archived", "deleted"]
    turns: list[SessionConversationTurnRead]


class AskSessionRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)


class AskSessionResponse(BaseModel):
    conversation: SessionConversationRead
    answer: SessionConversationTurnRead

File: frontend/src/adapters/synthesis.ts
import type {
  SessionConversation as SessionConversationApi,
  SessionReport as SessionReportApi,
  SessionTheme as SessionThemeApi,
  SessionThemeEvidence as SessionThemeEvidenceApi,
} from "@/api/types";
import type {
  SessionConversation,
  SessionReport,
  SessionTheme,
  ThemeEvidenceDetail,
} from "@/domain/types";

export function toThemeEvidence(value: SessionThemeEvidenceApi): ThemeEvidenceDetail {
  return {
    id: value.id,
    documentId: value.document_id,
    documentName: value.document_name,
    speaker: value.speaker,
    location: value.location,
    excerpt: value.excerpt,
    relevance: value.relevance,
    contextResultId: value.context_result_id,
  };
}

export function toSessionTheme(value: SessionThemeApi): SessionTheme {
  return {
    id: value.id,
    projectId: value.project_id,
    sessionId: value.session_id,
    name: value.name,
    summary: value.summary,
    status: value.status,
    confidence: value.confidence,
    sourceLabel: value.source_label,
    evidence: value.evidence.map(toThemeEvidence),
  };
}

export function toSessionReport(value: SessionReportApi): SessionReport {
  return {
    id: value.id,
    projectId: value.project_id,
    sessionId: value.session_id,
    status: value.status,
    sessionTitle: value.session_title,
    sessionType: value.session_type,
    sessionDate: value.session_date,
    durationMinutes: value.duration_minutes ?? undefined,
    participants: value.participants.map((participant) => ({
      id: participant.id,
      name: participant.name,
      role: participant.role ?? undefined,
      organization: participant.organization ?? undefined,
      notes: participant.notes ?? undefined,
    })),
    executiveSummary: value.executive_summary,
    items: value.items.map((item) => ({
      id: item.id,
      type: item.type,
      title: item.title,
      summary: item.summary,
      provenance: item.provenance,
      evidence: item.evidence.map(toThemeEvidence),
    })),
    detailedNotes: value.detailed_notes,
    generatedAt: value.generated_at,
  };
}

export function toSessionConversation(value: SessionConversationApi): SessionConversation {
  return {
    id: value.id,
    projectId: value.project_id,
    sessionId: value.session_id,
    status: value.status,
    turns: value.turns.map((turn) => ({
      id: turn.id,
      role: turn.role,
      content: turn.content,
      createdAt: turn.created_at,
      citations: turn.citations.map((citation) => ({
        id: citation.id,
        documentId: citation.document_id,
        documentName: citation.document_name,
        speaker: citation.speaker,
        location: citation.location,
        excerpt: citation.excerpt,
        contextResultId: citation.context_result_id,
      })),
    })),
  };
}
